openssl/src/crypto/dsa.rs

```diff
@@ -0,0 +1,351 @@
+                                              None, ptr::null_mut()))
+        };
+
+
+        try!(io::copy(&mut mem_bio, writer).map_err(StreamError));
+        Ok(())
+    }
+
+    /// Reads an DSA public key from PEM formatted data.
+    pub fn public_key_from_pem<R>(reader: &mut R) -> Result<DSA, SslError>
+        where R: Read
+    {
+        let mut mem_bio = try!(MemBio::new());
+        try!(io::copy(reader, &mut mem_bio).map_err(StreamError));
+
+        unsafe {
+            let dsa = try_ssl_null!(ffi::PEM_read_bio_DSA_PUBKEY(mem_bio.get_handle(),
+                                                                 ptr::null_mut(),
+                                                                 None,
+                                                                 ptr::null_mut()));
+            Ok(DSA(dsa))
+        }
+    }
+
+    /// Writes an DSA public key as PEM formatted data
+    pub fn public_key_to_pem<W>(&self, writer: &mut W) -> Result<(), SslError>
+        where W: Write
+    {
+        let mut mem_bio = try!(MemBio::new());
+
+        unsafe { try_ssl!(ffi::PEM_write_bio_DSA_PUBKEY(mem_bio.get_handle(), self.0)) };
+
+        try!(io::copy(&mut mem_bio, writer).map_err(StreamError));
+        Ok(())
+    }
+
+    pub fn size(&self) -> Result<isize, SslError> {
+        if self.has_q() {
+            unsafe { Ok(ffi::DSA_size(self.0) as isize) }
+        } else {
+            Err(SslError::OpenSslErrors(vec![]))
+        }
+    }
+
+    pub fn sign(&self, hash: hash::Type, message: &[u8]) -> Result<Vec<u8>, SslError> {
+        let k_len = try!(self.size()) as c_uint;
+        let mut sig = vec![0;k_len as usize];
+        let mut sig_len = k_len;
+        assert!(self.has_private_key());
+
+        unsafe {
+            try_ssl!(ffi::DSA_sign(hash.as_nid() as c_int,
+                                   message.as_ptr(),
+                                   message.len() as c_int,
+                                   sig.as_mut_ptr(),
+                                   &mut sig_len,
+                                   self.0));
+            sig.set_len(sig_len as usize);
+            sig.shrink_to_fit();
+            Ok(sig)
+        }
+    }
+
+    pub fn verify(&self, hash: hash::Type, message: &[u8], sig: &[u8]) -> Result<bool, SslError> {
+        unsafe {
+            let result = ffi::DSA_verify(hash.as_nid() as c_int,
+                                         message.as_ptr(),
+                                         message.len() as c_int,
+                                         sig.as_ptr(),
+                                         sig.len() as c_int,
+                                         self.0);
+
+            try_ssl_if!(result == -1);
+            Ok(result == 1)
+        }
+    }
+
+    pub fn as_ptr(&self) -> *mut ffi::DSA {
+        self.0
+    }
+
+    // The following getters are unsafe, since BigNum::new_from_ffi fails upon null pointers
+    pub fn p(&self) -> Result<BigNum, SslError> {
```

Comment:
It seems like we can coalesce `p` and `has_p` by just having this return a `Result<Option<BigNum>, SslError>`. This mirrors the existing RSA API, though, so that should probably happen in a follow up on the breaks branch.